clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -753,17 +755,20 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
                     log.trace("Adding fetched record for partition {} with offset {} to buffered record list", tp, position);
                     parsedRecords = new PartitionRecords<>(fetchOffset, tp, parsed);
                 } else if (!skippedRecords && (partition.records.sizeInBytes() > 0)) {
-                    // We did not read a single message from a non-empty buffer, because that message's size is
-                    // larger than the fetch size.
-                    Map<TopicPartition, Long> recordTooLargePartitions = Collections.singletonMap(tp, fetchOffset);
-                    throw new RecordTooLargeException("There are some messages at [Partition=Offset]: "
-                                                + recordTooLargePartitions
-                                                + " whose size is larger than the fetch size "
-                                                + this.fetchSize
-                                                + " and hence cannot be ever returned."
-                                                + " Increase the fetch size on the client (using max.partition.fetch.bytes),"
-                                                + " or decrease the maximum message size the broker will allow (using message.max.bytes).",
-                                                recordTooLargePartitions);
+                    if (completedFetch.responseVersion < 3) {
+                        // Implement the pre-KIP-74 behavior of throwing a RecordTooLargeException.
+                        Map<TopicPartition, Long> recordTooLargePartitions = Collections.singletonMap(tp, fetchOffset);
+                        throw new RecordTooLargeException("There are some messages at [Partition=Offset]: " +
+                                recordTooLargePartitions + " whose size is larger than the fetch size " + this.fetchSize +
+                                " and hence cannot be returned.  Please considering upgrading your broker to 0.10.0.1 or " +
+                                "newer to avoid this issue.  Alternately, increase the fetch size on the client (using " +
+                                ConsumerConfig.MAX_PARTITION_FETCH_BYTES_CONFIG + ")",
+                                recordTooLargePartitions);
+                    } else {
+                        // This should not happen with brokers that support KIP-74.
+                        throw new RuntimeException("Failed to make progress reading messages at " + tp + "=" +
```

Comment:
Can you add some more detail to this message? It would probably be pretty obscure for a user to see this pop up in the logs. Maybe we can mention that we received a non-empty fetch response from the server, but could not parse any valid records?